### ISSUE TITLE
Add patient lookup modal to dispensing page

### ISSUE THREAD

**Issue by wlthomson** [openmsupply/mobile#2541]
## Is your feature request related to a problem? Please describe.

As part of the patient API, a patient lookup modal needs to be added to the dispensing page.

## Describe the solution you'd like

When the action is dispatched for looking up a "non-local" patient, a modal will be displayed which allows the user to enter their patient details and query the server for results. Any matching results will be displayed in the list, and on selection will be added to the patient list in the dispensing page.  

## Implementation

- Add searchable modal similar to existing `AutocompleteSelector` component.
- Add button for "querying" (`AutocompleteSelector` query is dispatched by a debouncer on user input, which probably has too much overhead for querying an API).
- Will likely need to add a spinner in a separate issue after lookup functionality has been added.

## Describe alternatives you've considered

Another idea was to leave UI pretty much as is, but with additional logic to lookup from the API as well as locally (where the new action would be dispatched explicitly by the user, rather than via a debouncer on user input). Although this a bit nicer, it's a fair bit more work, and it would be tricky to make the UI work in an intuitive way without some sort of toggle between "local" and "non-local" patients.

## Additional context

See epic #2446.

**Comment by joshxg** [comment]
Just some thoughts:

> Add searchable modal similar to existing AutocompleteSelector component.

I imagine users are going to need to enter many details to find a particular patient and possibly search through a few too: i.e. might need to enter first, last, dob and address in order to get a probably list. I imagined using something similar to the current patient form (where you can enter any details into the fields) to be able to enter as many details as you like in the fields

> AutocompleteSelector query is dispatched by a debouncer on user input, which probably has too much overhead for querying an API

`AutoCompleteSelector` doesn't have an inbuilt debounce on input

> Will likely need to add a spinner in a separate issue after lookup functionality has been added.

There is a custom hook `useLoadingIndicator` which will pass a component the 'spinner' function

> UI work in an intuitive way without some sort of toggle between "local" and "non-local" patients

Not sure why you need to have a toggle for that?

**Comment by wlthomson** [comment]
> I imagine users are going to need to enter many details to find a particular patient and possibly search through a few too: i.e. might need to enter first, last, dob and address in order to get a probably list. I imagined using something similar to the current patient form (where you can enter any details into the fields) to be able to enter as many details as you like in the fields

I spent a bit of time thinking about this, and I mostly agree, but there are two main reasons why I decided on a modal for now:

- Time to develop, can always refactor to use the approach described above with pretty small cost.
- Consistency with the existing search. If we use that sort of query interface for looking up patients via the API, it seems difficult to justify why we shouldn't use it for local patients too. I.e. maybe it's better to keep the search interfaces consistent and update them both in a separate issue/feature?

`AutoCompleteSelector doesn't have an inbuilt debounce on input`

Are you sure? It uses `SearchBar` which has a default `debounceTimeout` of `250`?

> There is a custom hook `useLoadingIndicator` which will pass a component the 'spinner' function

Nice :+1:.

> > UI work in an intuitive way without some sort of toggle between "local" and "non-local" patients

This only applies to the alternative of using the existing dispensing page data table for the API lookup results. As local and non-local patents would be displayed in the same table, I think it would be confusing for the user as to how to "revert" the search from an API lookup back to querying only local patients, e.g.:

- user enters a query and clicks `"Query server"` (or something similar).
- the patient selection updates with a few patients retrieved from the server.
- the user realises they made a typo, and the patient is actually local, so now wants to adjust the typo and "revert" the selection back to local patients. 

The above situation would cause confusion, I think. Should the displayed list of patients retain all the patients queried from the API, or clear them when the search query is updated? What if the user then updates the query again and it now matches some of the "non-local" patients, should they display, or do we consider them "lost" once the query string is updated?

Basically, I think it's a bit tricky to handle and a separate UI is cleaner.

**Comment by joshxg** [comment]
I think maybe we're on different pages so I think I will just stay out of it with the last mention that I was assuming that after querying for patient/s the records were just integrated into the mobile database as per any other patient, except they're not editable

**Comment by wlthomson** [comment]
> I think maybe we're on different pages so I think I will just stay out of it with the last mention that I was assuming that after querying for patient/s the records were just integrated into the mobile database as per any other patient, except they're not editable

Yup, that's the plan. Like you said, I think maybe we are on different pages!

**Comment by wlthomson** [comment]
Going to go ahead and close this as covered by #2609.
